apps/prairielearn/src/lib/groups.ts

```diff
@@ -200,79 +265,96 @@ export async function joinGroup(
   const splitJoinCode = fullJoinCode.split('-');
   if (splitJoinCode.length !== 2 || splitJoinCode[1].length !== 4) {
     // the join code input by user is not valid (not in format of groupname+4-character)
-    flash('error', 'The join code has an incorrect format');
+    throw new Error('The join code has an incorrect format');
     return;
   }
 
   const groupName = splitJoinCode[0];
   const joinCode = splitJoinCode[1].toUpperCase();
 
-  // This is a best-effort check to produce a nice error message. Even if this
-  // fails due to a race condition, the `group_users_insert` sproc below will
-  // validate that the user isn't already in a group.
-  const existingGroupId = await getGroupId(assessmentId, userId);
-  if (existingGroupId != null) {
-    flash('error', 'You are already in another group.');
-    return;
-  }
-
   try {
-    await sqldb.callAsync('group_users_insert', [
-      assessmentId,
-      userId,
-      authnUserId,
-      groupName,
-      joinCode,
-    ]);
+    await sqldb.runInTransactionAsync(async () => {
+      const group = await sqldb.queryOptionalRow(
+        sql.select_group_by_name,
+        { group_name: groupName, assessment_id: assessmentId },
+        GroupSchema,
+      );
+      if (group == null || group.join_code !== joinCode) {
+        throw new Error('Group does not exist.');
+      }
+      await addUserToGroup(assessmentId, group.id, userId, authnUserId, true);
+    });
   } catch (err) {
-    flash(
-      'error',
-      `Failed to join the group with join code ${fullJoinCode}. It is already full or does not exist. Please try to join another one.`,
-    );
+    throw new Error(`Cannot join group "${fullJoinCode}": ${err.message}`);
```

Comment:
I generally dislike this style of wrapping errors since it completely obscures any information from the original error (`err`) when this is logged or processed by Sentry. I'd remove it for now; once we're running on a suitable version of Node, we can use [Error causes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/Error#cause).

Reply:
Replaced with a custom error class in 65872d4f3.